zulipterminal/ui_tools/buttons.py

```diff
@@ -31,28 +31,31 @@ def __init__(
         count: int = 0,
         count_style: Optional[str] = None,
     ) -> None:
-        if isinstance(prefix_character, tuple):
-            prefix = prefix_character[1]
-        else:
-            prefix = prefix_character
-        assert len(prefix) in (0, 1)
+        self.controller = controller
         self._caption = caption
+        self.show_function = show_function
         self.prefix_character = prefix_character
-        self.post_prefix_spacing = " " if prefix else ""
+        self.original_color = text_color
         self.count = count
         self.count_style = count_style
 
-        prefix_length = 0 if prefix == "" else 2
-        # Space either side, at least one space between
-        self.width_for_text_and_count = width - 3 - prefix_length
-
-        self.original_color = text_color
-        self.show_function = show_function
         super().__init__("")
-        self.update_count(count, text_color)
-        self.controller = controller
         urwid.connect_signal(self, "click", self.activate)
 
+        self.button_prefix = urwid.Text("")
+        self._w.contents[0] = (self.button_prefix, self._w.options("pack"))
+
+        self._label.set_wrap_mode("ellipsis")
+        self._label._cursor_position = 80
+        self._w.dividechars = 0
+
+        self.button_suffix = urwid.Text("")
+        self._w.contents[2] = (self.button_suffix, self._w.options("pack"))
+
+        self._w = urwid.AttrMap(self._w, None, "selected")
```

Comment:
Regarding the use of private attributes, I recognize that we already reset the internal widget in the existing implementation, and so it definitely seems less of an issue to set `self._w` once (since we do already) - and certainly an improvement to set up the widgets once here rather than on every update, since we can with this kind of approach.

I also recognize the benefit of reusing `self._label` to allow the reuse in the update method more easily and potentially allow setting the label elsewhere - eg. if we update stream, topic or user text.

What I'm less clear of is if it would be clearer to construct our own Columns widget in an AttrMap (using self._label), so we get precisely what we want more explicitly.

Reply:
Yes agreed, creating our own widget once and replacing _w once would also be cleaner (contents list would not be needed).